connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala

```diff
@@ -471,16 +481,120 @@ private class KeyValueGroupedDatasetImpl[K, V, IK, IV](
     }
   }
 
+  private def aggUntypedWithValueMapFunc(columns: TypedColumn[_, _]*): Dataset[_] = {
+    val originalDs = sparkSession.newDataset(ivEncoder, plan)
+
+    // Apply the value transformation, get a DS of two columns "iv" and "v".
+    // If any of "iv" or "v" consists of a single primitive field, wrap it with a struct so it
+    // would not be flattened.
```

Comment:
Again. I think flattening is not an issue. The aggregators are surprisingly lenient in that sense.